### PR TITLE
Remove some feature flag usage from libsyntax

### PR DESCRIPTION
**Pull request by erickt**
This removes the usage of `#[feature(into_cow, slice_patterns, box_syntax, box_patterns, quote, unsafe_destructor)]` from being used in libsyntax. My main desire for this is that it brings me one step closer to letting [syntex](https://github.com/erickt/rust-syntex) compile with stable rust. Hopefully this doesn't inconvenience rust development.

### CODE REVIEW

**Comment by erickt** [comment]
Fixed the tidy issue and eliminated a few other feature flags.

File: src/libsyntax/print/pp.rs [hunk 1]
```diff
@@ -265,8 +265,8 @@ pub fn mk_printer<'a>(out: Box<io::Write+'a>, linewidth: usize) -> Printer<'a> {
 /// In this implementation (following the paper, again) the SCAN process is
 /// the method called 'pretty_print', and the 'PRINT' process is the method
 /// called 'print'.
-pub struct Printer<'a> {
-    pub out: Box<io::Write+'a>,
+pub struct Printer<W> {
+    pub out: W,
```
Comment (alexcrichton) [inline src/libsyntax/print/pp.rs]: In the past we've had some pretty bad code bloat problems by having a totally generic pretty printer, so I'm curious what instigated this change?

Comment (erickt) [inline src/libsyntax/print/pp.rs]: I was trying to make it easy to get to the underlying `Vec` so I could get rid of the reference to `TraitObject` and the transmute, but I can try a different approach.

**Comment by erickt** [comment]
@alexcrichton: turned out I could have saved myself a bunch of time trying to get rid of `TraitObject` by just using a scope and a `&mut` :)

... I think serde might have gotten me addicted to type parameters...

**Comment by alexcrichton** [comment]
@bors: r+ 3721bdfb383e70b454aefe68bcc8b9295ee45744

Thanks!

**Comment by Manishearth** [comment]
```
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:838:13: 841:53 error: slice pattern syntax is experimental
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:838             [ast::TtToken(_, token::Ident(name_macro_rules, token::Plain)),
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:839              ast::TtToken(_, token::Not),
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:840              ast::TtToken(_, token::Ident(name_zip, token::Plain)),
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:841              ast::TtDelimited(_, ref macro_delimed)]
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:838:13: 841:53 help: add #![feature(slice_patterns)] to the crate attributes to enable
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:845:21: 847:62 error: slice pattern syntax is experimental
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:845                     [ast::TtDelimited(_, ref first_delimed),
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:846                      ast::TtToken(_, token::FatArrow),
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:847                      ast::TtDelimited(_, ref second_delimed)]
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:845:21: 847:62 help: add #![feature(slice_patterns)] to the crate attributes to enable
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:850:29: 851:80 error: slice pattern syntax is experimental
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:850                             [ast::TtToken(_, token::Dollar),
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:851                              ast::TtToken(_, token::Ident(name, token::Plain))]
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:850:29: 851:80 help: add #![feature(slice_patterns)] to the crate attributes to enable
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:857:29: 858:80 error: slice pattern syntax is experimental
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:857                             [ast::TtToken(_, token::Dollar),
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:858                              ast::TtToken(_, token::Ident(name, token::Plain))]
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/parse/mod.rs:857:29: 858:80 help: add #![feature(slice_patterns)] to the crate attributes to enable
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/ext/expand.rs:1966:13: 1966:16 error: slice pattern syntax is experimental
/home/ubuntu/src/rust-buildbot/slave/auto-linux-64-nopt-t/build/src/libsyntax/ext/expand.rs:1966             [b] => b,
```

**Comment by erickt** [comment]
@alexcrichton: I included a few more commits that haven't been reviewed. I don't think you've seen the last 4 yet.

**Comment by alexcrichton** [comment]
Looks good to me! Could this hold off on the last commit for now though? Duplicating that kind of code is a bit unfortunate and there's still other features in use by libsyntax so we can hold off on the duplication as long as possible (hopefully!)

**Comment by erickt** [comment]
@alexcrichton: Sure no problem. I'll remove that patch for now.

**Pull request merged**